src/client.ts

```diff
@@ -5297,38 +5298,44 @@ export class MatrixClient extends TypedEventEmitter<EmittedEvents, ClientEventHa
             ...res.events_before.map(mapper),
         ];
 
-        // Where the event is a thread reply (not a root) and running in MSC-enabled mode the Thread timeline only
-        // functions contiguously, so we have to jump through some hoops to get our target event in it.
-        // XXX: workaround for https://github.com/vector-im/element-meta/issues/150
-        if (Thread.hasServerSideSupport &&
-            this.supportsExperimentalThreads() &&
-            event.isRelation(THREAD_RELATION_TYPE.name)
-        ) {
-            const [, threadedEvents] = timelineSet.room.partitionThreadedEvents(events);
-            let thread = timelineSet.room.getThread(event.threadRootId);
-            if (!thread) {
-                thread = timelineSet.room.createThread(event.threadRootId, undefined, threadedEvents, true);
+        if (this.supportsExperimentalThreads()) {
+            const { threadId, shouldLiveInRoom } = timelineSet.room.eventShouldLiveIn(event);
+
+            if (!timelineSet.thread && !shouldLiveInRoom) {
+                // Thread response does not belong in this timelineSet
+                return undefined;
             }
 
-            const opts: IRelationsRequestOpts = {
-                direction: Direction.Backward,
-                limit: 50,
-            };
+            if (timelineSet.thread?.id !== threadId) {
+                // Event does not belong in this timelineSet
+                return undefined;
+            }
```

Comment:
I think this second conditional covers all of the cases:

 - If main timeline and thread event, return `undefined`
 - If thread timeline and main timeline event, return `undefined`
 - If thread timeline and event from different thread, return `undefined`
 - If the main timeline and main timeline event, return timeline
 - If thread timeline and event from same thread, return timeline
 
I feel like the first conditional doesn't add anything more? Am I missing something?